File: ImperatorToCK3Tests/ImperatorWorldTests/Provinces/ProvincesTests.cpp
#include "gtest/gtest.h"
#include <sstream>

#include "../ImperatorToCK3/Source/Imperator/Provinces/Provinces.h"
#include "../ImperatorToCK3/Source/Imperator/Provinces/Province.h"
#include "../ImperatorToCK3/Source/Imperator/Provinces/Pops.h"
#include "../ImperatorToCK3/Source/Imperator/Provinces/Pop.h"

TEST(ImperatorWorld_ProvincesTests, provincesDefaultToEmpty)
{
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "}";

	const Imperator::Provinces provinces(input);

	ASSERT_TRUE(provinces.getProvinces().empty());
}

TEST(ImperatorWorld_ProvincesTests, provincesCanBeLoaded)
{
	std::stringstream input;
	input << "=\n";
	input << "{\n";
	input << "42={}\n";
	input << "43={}\n";
	input << "}";

	const Imperator::Provinces provinces(input);
	const auto& provinceItr = provinces.getProvinces().find(42);
	const auto& provinceItr2 = provinces.getProvinces().find(43);

	ASSERT_EQ(42, provinceItr->first);
	ASSERT_EQ(42, provinceItr->second->getID());
	ASSERT_EQ(43, provinceItr2->first);
	ASSERT_EQ(43, provinceItr2->second->getID());
}

TEST(ImperatorWorld_ProvincesTests, popCanBeLinked)
{
	std::stringstream input;
	input << "={42={pop=8}}\n";
	Imperator::Provinces provinces(input);

	std::stringstream input2;
	input2 << "8={type=\"citizen\" culture=\"roman\" religion=\"paradoxian\"}\n";
	Imperator::Pops pops;
	pops.loadPops(input2);

	provinces.linkPops(pops);

	const auto& provinceItr = provinces.getProvinces().find(42);
	const auto& pop = provinceItr->second->getPops().find(8);

	ASSERT_TRUE(pop->second);
	ASSERT_EQ("citizen", pop->second->getType());
}

TEST(ImperatorWorld_ProvincesTests, multiplePopsCanBeLinked)
{
	std::stringstream input;
	input << "={\n";
	input << "43={ pop = 10}\n";
	input << "42={pop=8}\n";
	input << "44={pop= 9}\n";
	input << "}\n";
	Imperator::Provinces provinces(input);

	std::stringstream input2;
	input2 << "={\n";
	input2 << "8={type=\"citizen\" culture=\"roman\" religion=\"paradoxian\"}\n";
	input2 << "9={type=\"tribal\" culture=\"persian\" religion=\"gsg\"}\n";
	input2 << "10={type=\"freemen\" culture=\"greek\" religion=\"zoroastrian\"}\n";
	input2 << "}\n";
	Imperator::Pops pops;
	pops.loadPops(input2);

	provinces.linkPops(pops);

	const auto& provinceItr = provinces.getProvinces().find(42);
	const auto& pop = provinceItr->second->getPops().find(8);
	const auto& provinceItr2 = provinces.getProvinces().find(43);
	const auto& pop2 = provinceItr2->second->getPops().find(10);
	const auto& provinceItr3 = provinces.getProvinces().find(44);
	const auto& pop3 = provinceItr3->second->getPops().find(9);

	ASSERT_TRUE(pop->second);
	ASSERT_EQ("citizen", pop->second->getType());
	ASSERT_TRUE(pop2->second);
	ASSERT_EQ("freemen", pop2->second->getType());
	ASSERT_TRUE(pop3->second);
	ASSERT_EQ("tribal", pop3->second->getType());
}

TEST(ImperatorWorld_ProvincesTests, BrokenLinkAttemptThrowsWarning)
{
	std::stringstream input;
	input << "={\n";
	input << "42={ pop = 8 }\n";
	input << "44={ pop = 10 }\n"; /// no pop 10
	input << "}\n";
	Imperator::Provinces provinces(input);

	std::stringstream input2;
	input2 << "={\n";
	input2 << "8={type=\"citizen\" culture=\"roman\" religion=\"paradoxian\"}\n";
	input2 << "9={type=\"tribal\" culture=\"persian\" religion=\"gsg\"}\n";
	input2 << "}\n";
	Imperator::Pops pops;
	pops.loadPops(input2);

	std::stringstream log;
	auto* stdOutBuf = std::cout.rdbuf();
	std::cout.rdbuf(log.rdbuf());

	provinces.linkPops(pops);

	std::cout.rdbuf(stdOutBuf);
	auto stringLog = log.str();
	auto newLine = stringLog.find_first_of('\n');
	stringLog = stringLog.substr(0, newLine);

	ASSERT_EQ(" [WARNING] Pop ID: 10 has no definition!", stringLog);
}

File: ImperatorToCK3.UnitTests/Imperator/Provinces/ProvincesTests.cs
using commonItems;
using System;
using System.IO;
using Xunit;

namespace ImperatorToCK3.UnitTests.Imperator.Provinces {
	[Collection("Sequential")]
	[CollectionDefinition("Sequential", DisableParallelization = true)]
	public class ProvincesTests {
		[Fact]
		public void ProvincesDefaultToEmpty() {
			var reader = new BufferedReader("={}");
			var provinces = new ImperatorToCK3.Imperator.Provinces.ProvinceCollection(reader);

			Assert.Empty(provinces);
		}

		[Fact]
		public void ProvincesCanBeLoaded() {
			var reader = new BufferedReader(
				"=\n" +
				"{\n" +
				"42={}\n" +
				"43={}\n" +
				"}"
			);
			var provinces = new ImperatorToCK3.Imperator.Provinces.ProvinceCollection(reader);

			Assert.Equal((ulong)42, provinces[42].Id);
			Assert.Equal((ulong)43, provinces[43].Id);
		}

		[Fact]
		public void PopCanBeLinked() {
			var reader = new BufferedReader("={42={pop=8}}\n");
			var provinces = new ImperatorToCK3.Imperator.Provinces.ProvinceCollection(reader);

			var reader2 = new BufferedReader(
				 "8={type=\"citizen\" culture=\"roman\" religion=\"paradoxian\"}\n"
			);
			var pops = new ImperatorToCK3.Imperator.Pops.Pops();
			pops.LoadPops(reader2);
			provinces.LinkPops(pops);

			var province = provinces[42];
			var pop = province.Pops[8];

			Assert.NotNull(pop);
			Assert.Equal("citizen", pop.Type);
		}

		[Fact]
		public void MultiplePopsCanBeLinked() {
			var reader = new BufferedReader(
				"={\n" +
				"43={ pop = 10}\n" +
				"42={pop=8}\n" +
				"44={pop= 9}\n" +
				"}\n"
			);
			var provinces = new ImperatorToCK3.Imperator.Provinces.ProvinceCollection(reader);

			var reader2 = new BufferedReader(
				"={\n" +
				"8={type=\"citizen\" culture=\"roman\" religion=\"paradoxian\"}\n" +
				"9={type=\"tribal\" culture=\"persian\" religion=\"gsg\"}\n" +
				"10={type=\"freemen\" culture=\"greek\" religion=\"zoroastrian\"}\n" +
				"}\n"
			);
			var pops = new ImperatorToCK3.Imperator.Pops.Pops();
			pops.LoadPops(reader2);
			provinces.LinkPops(pops);

			var province = provinces[42];
			var pop = province.Pops[8];
			var province2 = provinces[43];
			var pop2 = province2.Pops[10];
			var province3 = provinces[44];
			var pop3 = province3.Pops[9];

			Assert.NotNull(pop);
			Assert.Equal("citizen", pop.Type);
			Assert.NotNull(pop2);
			Assert.Equal("freemen", pop2.Type);
			Assert.NotNull(pop3);
			Assert.Equal("tribal", pop3.Type);
		}

		[Fact]
		public void BrokenLinkAttemptThrowsWarning() {
			var reader = new BufferedReader(
				"={\n" +
				"42={ pop = 8 }\n" +
				"44={ pop = 10 }\n" + // no pop 10
				"}\n"
			);
			var provinces = new ImperatorToCK3.Imperator.Provinces.ProvinceCollection(reader);

			var reader2 = new BufferedReader(
				"={\n" +
				"8={type=\"citizen\" culture=\"roman\" religion=\"paradoxian\"}\n" +
				"9={type=\"tribal\" culture=\"persian\" religion=\"gsg\"}\n" +
				"}\n"
			);
			var pops = new ImperatorToCK3.Imperator.Pops.Pops();
			pops.LoadPops(reader2);

			var output = new StringWriter();
			Console.SetOut(output);

			provinces.LinkPops(pops);

			var logStr = output.ToString();
			Assert.Contains("[WARN] Pop with ID 10 has no definition!", logStr);
		}
	}
}
